cpp/src/spatial/hausdorff.cu

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required point_b_y applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+#include <type_traits>
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cuspatial/error.hpp>
+
+namespace {
+
+template<typename T>
+constexpr auto magnitude_squared(T a, T b) {
+    return a * a + b * b;
+}
+
+template <typename T>
+__global__ void kernel_hausdorff(int num_spaces,
+                                 T const* xs,
+                                 T const* ys,
+                                 cudf::size_type* space_offsets,
+                                 T* results)
+{
+    auto block_idx = blockIdx.y * gridDim.x + blockIdx.x;
+    auto num_results = num_spaces * num_spaces;
+
+    // each block processes a single result / pair of spaces
+    if (block_idx >= num_results)
+    {
+        return;
+    }
+
+    int space_a_idx   = block_idx % num_spaces;
+    int space_a_begin = space_a_idx == 0 ? 0 : space_offsets[space_a_idx - 1];
+    int space_a_end   =                        space_offsets[space_a_idx];
+
+    int space_b_idx   = block_idx / num_spaces;
+    int space_b_begin = space_b_idx == 0 ? 0 : space_offsets[space_b_idx - 1];
+    int space_b_end   =                        space_offsets[space_b_idx];
+
+    T min_dist_sqrd = 1e20;
+
+    int num_points_in_b = space_b_end - space_b_begin;
+
+    if (threadIdx.x < num_points_in_b)
+    {
+        T point_b_x = xs[space_b_begin + threadIdx.x];
+        T point_b_y = ys[space_b_begin + threadIdx.x];
+
+        for (int i = space_a_begin; i < space_a_end; i++)
+        {
+            T point_a_x = xs[i];
+            T point_a_y = ys[i];
+            T dist_sqrd = magnitude_squared(point_b_x - point_a_x, point_b_y - point_a_y);
+
+            min_dist_sqrd = min(min_dist_sqrd, dist_sqrd);
+        }
+    }
+
+    if (min_dist_sqrd > 1e10)
+    {
+        min_dist_sqrd = -1;
+    }
+
+    __shared__ T dist_sqrd[1024];
```

Comment:
The `1024` here is suppose to be the same as `NUM_THREADS` below, right? Maybe make it a `constexpr` global and use it here as well.